packages/lodestar/src/db/api/beacon/beacon.ts

```diff
@@ -102,4 +103,20 @@ export class BeaconDb extends DatabaseService implements IBeaconDb {
     return state.validators.findIndex(value => this.config.types.BLSPubkey.equals(value.pubkey, publicKey));
   }
 
+  public async getStateForSlot(slot: Slot): Promise<BeaconState> {
+    const epoch = computeEpochAtSlot(this.config, slot);
+    const startSlot = computeStartSlotAtEpoch(this.config, epoch);
+    const blockRoot = await this.chain.getBlockRoot(startSlot);
+    if (blockRoot) {
+      const block = await this.block.get(blockRoot);
+      return await this.state.get(block.message.stateRoot.valueOf() as Uint8Array);
+    }
+    const headBlock = await this.block.getChainHead();
+    const state = await this.state.get(headBlock.message.stateRoot.valueOf() as Uint8Array);
+    if(state.slot < startSlot) {
+      processSlots(this.config, state, startSlot);
```

Comment:
is this good to put `processSlots` here in db? cannot find a better place ...

Reply:
I think using `processSlots` is fine, though we should be processing slots up to the requested slot.
We're trying to get the state at the slot in question to verify the block signature, and the proposer gets changed per-slot, so I believe we need the _exact_ slot.

One thing to keep in mind is this part of the database code will soon be changed. (Soon we'll be keeping recent states in memory).

An alternative to doing the `processSlots` here, you might return the first available state at or before the requested slot, and make the consumer (ie: the gossip validator) run `processSlots`.

Reply:
yeah I get your logical points, I can do that but still have concerns regarding having to do `processSlots` up until the requested slot (may be wrong but here's what I see from the specs). To get correct state to validate block proposer signature, we need 3 things: the seed, the slot(not a concern) and active validator indixes

+ the seed is calculated from `randao_mixes` of 2 epochs earlier. If our head is older than that, just return null and make the validation failed. Running `processSlots` does not give correct `randao_mixes` for the current epoch because it should be updated block by block (in `process_block`).
+ From the 1st look, running `process_slots` may help get correct active validator indexes because we run thru `process_epoch` then `processRegistryUpdates`. But a validator can only becomes active in `epoch + 1 + config.params.MAX_SEED_LOOKAHEAD` which is `epoch+5`. In most of the cases, the head state should have same active validator indexes. If not, running `process_slots` does not help because the head is 5 epochs older than the block and it failed because of the `seed` condition above.

My thinking is that it's not wrong to run `processSlots` here, it's just not necessary to run it at all because we have same result anyway.

Reply:
> a validator can only becomes active in epoch + 1 + config.params.MAX_SEED_LOOKAHEAD

If you have validator that deposited some time ago and it's suppose to become active in current epoch +1. Running process slots till current epoch + 1 will change active validator indices

Reply:
Maybe it's not like that, let's look into the detail of `processRegistryUpdates` (lmk if I look at wrong place):
```
activationQueue.slice(0, getValidatorChurnLimit(config, state)).forEach((validator) => {
    if (validator.activationEpoch === FAR_FUTURE_EPOCH) {
      validator.activationEpoch = computeActivationExitEpoch(config, currentEpoch);
    }
  });
```
and 
```
export function computeActivationExitEpoch(config: IBeaconConfig, epoch: Epoch): Epoch {
  return epoch + 1 + config.params.MAX_SEED_LOOKAHEAD;
}
```
Compared head state and slot, we always have same set of active validators unless head state is more than 4/5 epochs (MAX_SEED_LOOKAHEAD) earlier compared to slot. And in that case `randao_mixes` is not correct anyway if we run `processSlots` as analyzed above. Hence with or without running `processSlots`, we both have the result as `false`.

Reply:
You are looking at right place but interpreting code wrong.
Scenario:
- we are at epoch 10
- validator deposited, passed activation queue and now validator.activationEpoch = 11(+1 + MAX_SEED_LOOKAHEAD)
- we received block in epoch 11
- we process slots till epoch 11 and on epoch transition that validator is not active and changes active validator indices

Reply:
In my understanding, in your scenario we don't need to run `processSlots()`. If you look at my validation, I'll pick head state and use `{...state, slot}` to verify the block signature. It should get the correct active validators.

On the other hand in `processSlots => processRegistryUpdates`, it only update `activationEpoch` (normally `current epoch+5`) while the place to get active validators is decided in `getActiveValidatorIndices`. As long as we put in the correct slot to state, it should get correct active valiator index.

Reply:
If I look at your explanation and mine, I think the main confusion comes from using `processSlots()` as yours/Cayman's and using `{...headState,slot}` as mine (I studied this way in `getEpochProposers`). I think using `{..headState,slot}` should be enough.

Reply:
process_epoch also modifes randao_mixes in final updates. Anyways, if we don't do process_slots, we are exposing ourselves to bugs caused by change in spec which affects proposers and it's way too hard to determine what will or will not affect it.

Reply:
> process_epoch also modifes randao_mixes in final updates
this `randao_mixes` needs to be "mixed"/updated by `process_block` of every block of epoch (then `process_randao`). If we merely run `process_slots` then the resulting `randao_mixes` may not be correct in final update.

> Anyways, if we don't do process_slots, we are exposing ourselves to bugs caused by change in spec which affects proposers and it's way too hard to determine what will or will not affect it.
For this point, I agree with this. I wish it's part of the spec so that all client implementation can follow.

Let's go with `process_slots`, sorry took your time lol ..